File: Data Structures Algorithns/Old/Stack.cpp
#include <iostream>

class Node
{
public:
    Node *next;
    void *data;
    Node(void *data)
    {
        this->data = data;
        this->next = NULL;
    }
    ~Node();
};

class Stack
{
private:
    Node *start;
    int count;

public:
    Stack()
    {
        this->start = NULL;
        this->count = 0;
    }
    void push(void *data)
    {
        Node *node = new Node(data);
        if (this->start)
        {
            Node *current = this->start;
            this->start = node;
            node->next = current;
        }
        else
            this->start = node;
        this->count++;
    }
    void *peek()
    {
        return this->start->data;
    }
    void *pop()
    {
        Node *node = this->start;
        this->start = this->start->next;
        this->count--;
        return node->data;
    }
    int getCount() { return this->count; }
};

int main(int argc, char const *argv[])
{
    Stack s;
    int a = 1;
    int b = 2;
    ;
    int c = 3;
    s.push(&a);
    s.push(&b);
    s.push(&c);

    std::cout << *(int *)(s.peek()) << std::endl;
    std::cout << *(int *)(s.pop()) << std::endl;
    std::cout << *(int *)(s.pop()) << std::endl;
    std::cout << *(int *)(s.pop()) << std::endl;
    return 0;
}

File: Data Structures Algorithns/Old/Stack.py
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None
    def __str__(self):
        return str(self.data)
        
class Stack:
    def __init__(self):
        self.top = None
        self.count = 0
    
    def push(self, data):
        node = Node(data)
        if self.top:
            node.next = self.top
        self.top = node
        self.count+=1
    
    def peek(self):
        return self.top
        
    def pop(self):
        node = self.top
        self.top = self.top.next
        return node

s = Stack()
s.push(1)
s.push(2)
s.push(3)
print(s.peek())
print(s.pop())
print(s.pop())
print(s.pop())        